pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -63,21 +85,60 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
-	// TODO: call pre-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
-	followup, err := next.Execute(msg)
+
+	// trigger the event if the message type is registered (if it's not outbound)
+	eventsTriggered := false
+	if !msg.Outbound {
+		eventsTriggered, err = s.sendEvent(msg, thid, next)
+		if err != nil {
+			return fmt.Errorf("send events failed: %w", err)
+		}
+	}
+
+	// if no events are trigger continue the execution
+	if !eventsTriggered {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// Register events on DID Exchange protocol.
+func (s *Service) Register(msgType string, ch chan<- event.DIDCommEvent) error {
+	if msgType != ConnectionInvite && msgType != ConnectionRequest &&
+		msgType != ConnectionResponse && msgType != ConnectionAck {
+		return fmt.Errorf("event registration not supported for the message type %s", msgType)
+	}
+
+	if _, ok := s.events[msgType]; ok {
+		s.events[msgType] = append(s.events[msgType], ch)
```

Comment:
I think that you'll need a mutex around s.events.

Reply:
[Added](https://github.com/hyperledger/aries-framework-go/compare/2fc8ca5f292de18ad5657f1f1317e24252476a3c..9e40655340de2e0f5c4fc9f2074c4f8cd3e1126d#diff-f761c91e753eba1c42a68076df06ce12R61) mutex.